### ISSUE TITLE
[jvm-packages] eval_set for xgboost4j-spark

### ISSUE THREAD

**Issue by frenzykryger** [dmlc/xgboost#3231]
There is no way to set custom evaluation set for `ml.dmlc.xgboost4j.scala.spark.XGBoost#trainDistributed`. Code inside uses private `ml.dmlc.xgboost4j.scala.spark.Watches` class which just splits train with predefined trainTestRatio and doesn't accept any custom eval set through params.
Is there any particular reason for this limitation or it's just stub and can be extended for example with DMatrix passed through params? Is there any complications caused by fact that this is distributed XGBoost? How such dataset should be stored in params then, as DMatrix or RDD, or something else?

**Comment by CodingCat** [comment]
I think there is a comment when bring the code in, https://github.com/dmlc/xgboost/pull/2710#discussion_r141479583

Would you like to give this requirement a shot?

**Comment by hcho3** [comment]
All feature requests are now consolidated to #3439. A new issue should be opened if someone decides to actively work on implementing this feature.

**Comment by CodingCat** [comment]
I will work on eval set this week

**Comment by hcho3** [comment]
Consolidating to the feature request tracker #3439. Feel free to re-open this issue when anyone starts working on this.

**Comment by CodingCat** [comment]
the feature is implemented in https://github.com/dmlc/xgboost/pull/3910

**Comment by eliyara** [comment]
Hi @CodingCat,

I need to define a separate validation set for cross validation, using xgboost4j on spark. I tried the approach [here](https://github.com/CodingCat/xgboost/blob/c4fa8d369950282f8a2f6e8c6cbd725e403b0219/jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/spark/SparkTraining.scala). It does not look that it setting validation sets make any difference! Should I expect the following set up work as cross validation does? 

```
        val params = scala.collection.mutable.Map(
            "eta" -> 0.1,
            "objective" -> "binary:logistic",
            "eval_sets" -> Map("dev" -> dev_df))
        val booster = new XGBoostClassifier(params.toMap)
        booster.setFeaturesCol("features")
        booster.setLabelCol("label")
        booster.setMaxDepth(5)
        booster.setNumRound(150)
        booster.setNumWorkers(4)
        val xgb_model = booster.fit(train_df)
```
